File: tap-ts/src/message-objects/authorize.ts
import { TAPAgent } from "../agent";
import { BaseMessage } from "./base-message";

/**
 * AuthorizeMessage - Represents a TAP Authorize message
 */
export class AuthorizeMessage extends BaseMessage {
  /**
   * Create a new authorize message
   */
  constructor(agent: TAPAgent, message: any) {
    super(agent, message);
  }

  /**
   * Get the reason for the authorization
   */
  get reason(): string | undefined {
    return this.body.reason;
  }

  /**
   * Set the reason for the authorization
   */
  setReason(reason: string): this {
    this.body.reason = reason;
    return this;
  }

  /**
   * Get the settlement address for the authorization
   */
  get settlementAddress(): string | undefined {
    return this.body.settlementAddress;
  }

  /**
   * Set the settlement address for the authorization
   */
  setSettlementAddress(settlementAddress: string): this {
    this.body.settlementAddress = settlementAddress;
    return this;
  }

  /**
   * Get the expiry timestamp for the authorization
   */
  get expiry(): string | undefined {
    return this.body.expiry;
  }

  /**
   * Set the expiry timestamp for the authorization
   */
  setExpiry(expiry: string): this {
    this.body.expiry = expiry;
    return this;
  }
}

File: tap-msg/src/message/authorize.rs
//! Authorize message type for the Transaction Authorization Protocol.
//!
//! This module defines the Authorize message type, which is used
//! for authorizing transactions in the TAP protocol.

use serde::{Deserialize, Serialize};

use crate::error::{Error, Result};
use crate::TapMessage;

/// Authorize message body (TAIP-4).
#[derive(Debug, Clone, Serialize, Deserialize, TapMessage)]
#[tap(message_type = "https://tap.rsvp/schema/1.0#Authorize")]
pub struct Authorize {
    /// ID of the transaction being authorized.
    #[tap(thread_id)]
    pub transaction_id: String,

    /// Optional settlement address in CAIP-10 format.
    /// Required when sent by a VASP representing the beneficiary unless the original
    /// request contains an agent with the settlementAddress role.
    #[serde(rename = "settlementAddress", skip_serializing_if = "Option::is_none")]
    pub settlement_address: Option<String>,

    /// Optional expiry timestamp in ISO 8601 format.
    /// After this time, if settlement has not occurred, the authorization should be
    /// considered invalid and settlement should not proceed.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub expiry: Option<String>,
}

impl Authorize {
    /// Create a new Authorize message
    pub fn new(transaction_id: &str) -> Self {
        Self {
            transaction_id: transaction_id.to_string(),
            settlement_address: None,
            expiry: None,
        }
    }

    /// Create a new Authorize message with a settlement address
    pub fn with_settlement_address(transaction_id: &str, settlement_address: &str) -> Self {
        Self {
            transaction_id: transaction_id.to_string(),
            settlement_address: Some(settlement_address.to_string()),
            expiry: None,
        }
    }

    /// Create a new Authorize message with all optional fields
    pub fn with_all(
        transaction_id: &str,
        settlement_address: Option<&str>,
        expiry: Option<&str>,
    ) -> Self {
        Self {
            transaction_id: transaction_id.to_string(),
            settlement_address: settlement_address.map(|s| s.to_string()),
            expiry: expiry.map(|s| s.to_string()),
        }
    }
}

impl Authorize {
    /// Custom validation for Authorize messages
    pub fn validate_authorize(&self) -> Result<()> {
        if self.transaction_id.is_empty() {
            return Err(Error::Validation(
                "Transaction ID is required in Authorize".to_string(),
            ));
        }

        Ok(())
    }
}
